nomad/acl.go

```diff
@@ -2,15 +2,117 @@ package nomad
 
 import (
 	"fmt"
+	"net"
 	"time"
 
 	metrics "github.com/armon/go-metrics"
 	lru "github.com/hashicorp/golang-lru"
 	"github.com/hashicorp/nomad/acl"
+	"github.com/hashicorp/nomad/helper"
 	"github.com/hashicorp/nomad/nomad/state"
 	"github.com/hashicorp/nomad/nomad/structs"
 )
 
+// Authenticate extracts an AuthenticatedIdentity from the request context or
+// provided token. The caller can extract an acl.ACL, WorkloadIdentity, or other
+// identifying token to use for authorization.
+func (s *Server) Authenticate(ctx *RPCContext, secretID string) (*structs.AuthenticatedIdentity, error) {
+	// get the user ACLToken or anonymous token
+	aclToken, err := s.ResolveSecretToken(secretID)
+	identity := &structs.AuthenticatedIdentity{ACLToken: aclToken}
+
+	switch err {
+	case nil:
+		// missing or anonymous ACL will fall-through for further annotation
+		// based on the RPC context
+		if aclToken == nil || aclToken != structs.AnonymousACLToken {
+			return identity, nil
+		}
+
+	case structs.ErrTokenExpired:
+		return nil, err
+
+	case structs.ErrTokenInvalid:
+		// if it's not a UUID it might be an identity claim
+		claims, err := s.VerifyClaim(secretID)
+		if err != nil {
+			return nil, err
+		}
+		// unlike the state queries, errors here are invalid tokens
```

Comment:
Is this really saying, "we tried to read the ACL from state store and it was invalid, now we tried to read it as an identity claim and it was invalid, and now there are no other options." ?

Reply:
Yes! Will clarify.

Reply:
Also, the comment should be on the error above, not this line 🤦